Listen to private data in the events

arsulegai opened the pull request:
Listen to the private data map from the deliver response message.
Addresses #153

arsulegai committed:
Listen to private data in the events

Signed-off-by: S m, Aruna <aruna.mohan@walmart.com>

bestbeforetoday reviewed (changes requested):
This approach looks generally OK to me. Please consider the specific review comments I have added an also:

1. The BlockInfo class is a mashup of two (and now three) different behaviours depending on the event type it represents. This looks to me like BlockInfo should be an interface with three distinct (package private) implementation classes: FilteredBlockInfo, FullBlockInfo, PrivateBlockInfo (or if you prefer maybe BlockWithPrivateDataInfo). BlockInfo just describes the available methods and each implementation class holds only the logic for its block type and we lose all the convoluted if/else method implementations. This would mean BlockEvent should probably also be an interface (extending BlockInfo) with two distinct implementation classes: FullBlockEvent and PrivateBlockEvent. A factory method (on BlockEvent) can be used by the PeerEventServiceClient to create a suitable implementation instance based on the DeliverResponse type.
2. Some tests are needed. I appreciate this code is not structured to make unit testing easy but you could modify or extend the existing integration testing of events to use the new private type.

arsulegai commented on src/main/java/org/hyperledger/fabric/sdk/Channel.java:
```diff
@@ -1147,7 +1147,7 @@ public PeerOptions getPeersOptions(Peer peer) {
      */
 
-    PeerOptions setPeerOptions(Peer peer, PeerOptions peerOptions) throws InvalidArgumentException {
+    public PeerOptions setPeerOptions(Peer peer, PeerOptions peerOptions) throws InvalidArgumentException {
```
Given that the peer options is a way to let the SDK request for private data. If one uses the SDK, they should have an ability to set the option. Is there a concern with this exposure?

Alternatives:
1. Consider a new method to set the peer's block request type option explicitly.
2. Consider a global setting at the channel level (instead of a peer option), to request generically private data from all the peers.
Let me know your thoughts.

bestbeforetoday replied:
I’m not sure it is the way I would have designed the API but the suggested usage pattern can be seen here:

- https://github.com/hyperledger/fabric-sdk-java/blob/9c4c3aa075b8c5673dc997721e51db2dfd70d024/src/test/java/org/hyperledger/fabric/sdkintegration/End2endAndBackAgainIT.java#L651
- https://github.com/hyperledger/fabric-sdk-java/blob/9c4c3aa075b8c5673dc997721e51db2dfd70d024/src/test/java/org/hyperledger/fabric/sdkintegration/UpdateChannelIT.java#L489

arsulegai replied:
Thanks for the test case references. Otherwise, I was heading in the direction to reuse the channel instance created using the gateway SDK.

arsulegai commented:
Thanks @bestbeforetoday for the detailed review. These comments shows an encouraging sign and vindicates the approach in general. I will get the interfaces/enum added necessary to make the code better readable/cleaner.

bestbeforetoday commented on src/main/java/org/hyperledger/fabric/sdk/BlockInfo.java:
```diff
@@ -31,6 +31,7 @@
 import org.hyperledger.fabric.protos.peer.TransactionPackage;
 import org.hyperledger.fabric.sdk.exception.InvalidProtocolBufferRuntimeException;
 import org.hyperledger.fabric.sdk.transaction.ProtoUtils;
+import sun.jvm.hotspot.utilities.AssertionFailure;
```
This is failing the build and will prevent it running on non-Oracle/Sun JVMs, so this import can be used

arsulegai replied:
Removed. Error on my end, ignored the auto-IDE imports during commit.

arsulegai commented on src/main/java/org/hyperledger/fabric/sdk/Channel.java:
```diff
@@ -6416,11 +6416,12 @@ protected void finalize() throws Throwable {
         protected Long startEvents;
         protected Long stopEvents = Long.MAX_VALUE;
         protected boolean registerEventsForFilteredBlocks = false;
+        protected boolean registerEventsForPrivateData = false;
```
Agree

arsulegai commented on src/main/java/org/hyperledger/fabric/sdk/Channel.java:
```diff
@@ -6451,17 +6452,39 @@ public boolean isRegisterEventsForFilteredBlocks() {
+         *
+         * @return true if private data blocks will be returned by the peer eventing service.
+         */
+        public boolean isRegisterEventsForPrivateData() {
```
Sure, agree.

arsulegai commented on src/main/java/org/hyperledger/fabric/sdk/PeerEventServiceClient.java:
```diff
@@ -59,6 +59,7 @@
 
     private final PeerOptions peerOptions;
     private final boolean filterBlock;
+    private final boolean blockAndPrivateData;
```
Sure, agree.

arsulegai commented on src/main/java/org/hyperledger/fabric/sdk/PeerEventServiceClient.java:
```diff
@@ -303,7 +309,8 @@ public void onCompleted() {
             };
 
-            nso = filterBlock ? broadcast.deliverFiltered(so) : broadcast.deliver(so);
+            nso = filterBlock ? broadcast.deliverFiltered(so) :
+                    blockAndPrivateData ? broadcast.deliverWithPrivateData(so) : broadcast.deliver(so);
```
Moved to switch on event type enum.

arsulegai committed:
Add integration tests for the new PDC events feature

Signed-off-by: S m, Aruna <aruna.mohan@walmart.com>

arsulegai commented:
> This approach looks generally OK to me. Please consider the specific review comments I have added an also:
> 
> 1. The BlockInfo class is a mashup of two (and now three) different behaviours depending on the event type it represents. This looks to me like BlockInfo should be an interface with three distinct (package private) implementation classes: FilteredBlockInfo, FullBlockInfo, PrivateBlockInfo (or if you prefer maybe BlockWithPrivateDataInfo). BlockInfo just describes the available methods and each implementation class holds only the logic for its block type and we lose all the convoluted if/else method implementations. This would mean BlockEvent should probably also be an interface (extending BlockInfo) with two distinct implementation classes: FullBlockEvent and PrivateBlockEvent. A factory method (on BlockEvent) can be used by the PeerEventServiceClient to create a suitable implementation instance based on the DeliverResponse type.
> 2. Some tests are needed. I appreciate this code is not structured to make unit testing easy but you could modify or extend the existing integration testing of events to use the new private type.

Thanks for the feedback. I will continue refactoring for the point 1 in upcoming PRs. But extended an existing IT similar to other eventing test case for the private data.

bestbeforetoday committed:
Refactor private data events

- Expose enum for block event type instead of multiple boolean getters to determine event type.

Signed-off-by: Mark S. Lewis <mark_lewis@uk.ibm.com>

bestbeforetoday reviewed (commented):
This change looks pretty good. Thank you for taking the time to rework it.

I had a look at whether it could be refactored to use inheritance or a strategy pattern to provide different implementations based on block event type instead of the cumbersome tests of type and different implementations within each method, but the code in this area looks to be so tangled that this would be a much larger piece of work. So I am happy with this pragmatic approach.

The one change I think might still be worth making is to avoid adding an `isBlockAndPrivate()` method, leaving the caller to figure out the block type by calling some combination of the `is...()` methods, and just go ahead and expose a type enum with a `getType()` method. I've made this change in the last commit on [this branch](https://github.com/bestbeforetoday/fabric-sdk-java/commits/pdc-events) in my fork. Have a look and, if it looks reasonable to you, go ahead and include that change in your PR so I can merge it.

arsulegai commented:
@bestbeforetoday done. Cherry-picked your commit, thanks for helping out. I didn't want to assume why an exception was thrown in earlier case, thought of not breaking much.

On the first point to make this part of code more readable and to use better programming constructs, I am on it and will raise more PRs soon. It's taking some time and didn't want to block the pdc eventing capability until then.

It would also help if we can have a patch release with this feature.

bestbeforetoday commented:
This is published in [v2.2.12](https://github.com/hyperledger/fabric-sdk-java/releases/tag/v2.2.12). Many thanks for the contribution!
